### Explorando e Analisando dados das vendas de carro pelo eBay na Alemanha

O objetivo desse projeto é realizar limpar os dados e realizar uma análise do dataset originado do scraping da seção de classificados do eBay alemão

O dataset original havia sido colocado inicialmente no Kaggle mas não se encontra mais lá, podendo ser encontrado [aqui](https://data.world/data-society/used-cars-data)

O dataset utilizado por essa análise é uma amostra da Dataquest que utiliza apenas 50000 pontos além de terem sujado um pouco mais os dados.

O dicionário do dataset dá as seguintes informações:

* dateCrawled - A data de quando ocorreu o scrap dessa informação pela primeira vez
* name - Nome do carro
* seller - Se o vendedor é pessoa física ou comercial
* offerType - O tipo da venda
* price - O preço do carro
* abtest - Se a listagem está incluída em um teste A/B
* vehicleType - Tipo do veículo
* yearOfRegistration - O ano em que o carro foi registrado
* gearbox - Tipo de transmissão
* powerPS - Potência do carro em PS
* model - Modleo do carro
* kilometer - Kilometragem do carro
* fuelType - Qual o combustível que o carro utiliza
* brand - Marca do carro
* notRepairedDamage - Se o carro possui algum dano não consertado
* dateCreated - Quando o anúncio do carro foi criado no eBay
* nrOfPictures - Quantidade de fotos no anúncio
* postalCode - Código postal da localização do veículo
* lastSeenOnline - Última vez que o crawler viu o anúncio online


In [1]:
# Importando as bibliotecas a serem utilizadas
import pandas as pd
import numpy as np

In [2]:
# Lendo os dados
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Após um primeira visão inicial podemos ver que nas colunas vehicleType, gearbox, model, fuelType e notRepairedDamage temos alguns dados faltantes

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# Renomeando as colunas utilizando snake case, e modificando alguns nomes
# para ficarem com nomes mais descritivos
autos.rename(columns = {'dateCrawled': 'date_crawled', 'offerType': 'offer_type', 
       'vehicleType': 'vehicle_type', 'yearOfRegistration':'registration_year', 'powerPS':'power_ps', 
       'monthOfRegistration':'registration_month', 'fuelType':'fuel_type', 
       'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}, inplace = True)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [9]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Como podemos ver acima e também na tabela resultante do método info, as colunas price e odometer estão como texto. Iremos converter esses dados para númerico para podems realizar mais análises

Além disso, chama a atenção que as colunas seller e offer_type tem, majoritariamente o mesmo valor, já que em uma dataset com 50000 linhas, ambos apresentam 49999 com o mesmo valor.

Também iremos investigar um pouco mais essas colunas

In [10]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [12]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','')
autos['odometer'] = autos['odometer'].astype(float)

In [13]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','')
autos['price'] = autos['price'].astype(float)

C:\Anaconda\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  """Entry point for launching an IPython kernel.


In [14]:
autos.rename(columns = {'odometer':'odometer_km'}, inplace = True)

In [15]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Com os dados das colunas price e odometer podemos começar a analisar essas colunas

In [16]:
autos['price'].unique().shape

(2357,)

In [17]:
autos['price'].describe().apply("{0:.2f}".format)

count       50000.00
mean         9840.04
std        481104.38
min             0.00
25%          1100.00
50%          2950.00
75%          7200.00
max      99999999.00
Name: price, dtype: object

Podemos ver que tanto o preço mínimo quanto máximo chamam a atençaõ por serem irreais. Prática comum em sites como eBay para chamar atenção do comprador em potencial. Normalmente eles anunciam o preço verdadeiro na descrição do anúncio, para a análise aqui eles não nos importam

In [18]:
autos['price'].loc[autos['price'] < 10]

27       0.0
55       1.0
71       0.0
80       0.0
87       0.0
        ... 
49884    0.0
49943    0.0
49960    0.0
49974    0.0
49984    0.0
Name: price, Length: 1585, dtype: float64

Preços muito pequenos podem parecer incomuns, mas como o eBay possue uma opção de leilões, há a possibilidade haver anúncios começando com valores bem baixos, snedo assim, iremos mantê-los

In [19]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

In [20]:
autos['price'].sort_values(ascending = False).head(30)

39705    99999999.0
42221    27322222.0
27371    12345678.0
39377    12345678.0
47598    12345678.0
2897     11111111.0
24384    11111111.0
11137    10000000.0
47634     3890000.0
7814      1300000.0
22947     1234566.0
43049      999999.0
514        999999.0
37585      999990.0
36818      350000.0
14715      345000.0
34723      299000.0
35923      295000.0
12682      265000.0
47337      259000.0
38299      250000.0
37840      220000.0
40918      198000.0
43668      197000.0
28090      194000.0
20351      190000.0
17140      180000.0
11433      175000.0
32840      169999.0
18509      169000.0
Name: price, dtype: float64

In [21]:
autos.loc[autos['price'] > 350000,]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

            price   abtest vehicle_type  registration_year    gearbox  \
514      999999.0     test        kombi               2009    manuell   
2897   11111111.0     test    limousine               1973    manuell   
7814    1300000.0  control        coupe               1992        NaN   
11137  10000000.0  control        coupe               1960    manuell   
22947   1234566.0  control        kombi               1999  automatik   
24384  11111111.0     test          NaN               1995        NaN   
27371  12345678.0  control          NaN               2017        NaN   
37585    999990.0     test    limousine               1985    manuell   
39377  12345678.0  control          NaN               2018    manuell   
39705  99999999.0  control    limousine               1999  automatik   
42221  27322222.0  control    limousine               2014    manuell   
43049    999999.0     test          bus               1981    manuell   
47598  12345678.0  control    limousine               2001    manuell   
47634   3890000.0     test        coupe               2006        NaN   

       power_ps        model  odometer_km  registration_month fuel_type  \
514         101        focus     125000.0                   4    benzin   
2897         48       escort      50000.0                   3    benzin   
7814          0          NaN      50000.0                  12       NaN   
11137       368          NaN     100000.0                   1    benzin   
22947       190          NaN     150000.0                   2    diesel   
24384         0          NaN     150000.0                   0       NaN   
27371        95        punto     150000.0                   0       NaN   
37585       111        jetta     150000.0                  12    benzin   
39377        95          v40     150000.0                   6       NaN   
39705       224     s_klasse     150000.0                   9    benzin   
42221       163           c4      40000.0                   2    diesel   
43049        70  transporter     150000.0                   1    benzin   
47598       101       vectra     150000.0                   3    benzin   
47634       799          NaN       5000.0                   7       NaN   

                brand unrepaired_damage           ad_created  postal_code  \
514              ford              nein  2016-03-17 00:00:00        12205   
2897             ford              nein  2016-03-12 00:00:00        94469   
7814   sonstige_autos              nein  2016-04-04 00:00:00        60598   
11137  sonstige_autos              nein  2016-03-29 00:00:00        73033   
22947             bmw               NaN  2016-03-22 00:00:00        17454   
24384      volkswagen               NaN  2016-03-21 00:00:00        18519   
27371            fiat               N

Podemos ver que há preços que fazem sentidos, como os da Ferrari contudo, mas muitos preços são arbitrários, como sequencia 123456 ou 999999 para carros que não tem esse valor, estratégia utilizada apenas para chamar atenção. Sendo assim, o cutoff será no 350000

In [22]:
autos = autos.loc[autos['price'] <=350000]

In [23]:
autos.price.describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

Interessante notar que saímos de uma média de quase 10 mil doláres para algo por volta de 6 mil doláres

In [24]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

In [25]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Impossível o máximo ser 9999 e o min ser 1000

In [26]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [27]:
autos[['name','registration_year']].loc[(autos['registration_year'] < 1930)]

name  registration_year
3679                                         Suche_Auto               1910
10556                                        UNFAL_Auto               1800
15898       Tausch_alles_aus_meinen_Anzeigen_gegen_Auto               1910
21416                           Essex_super_six__Ford_A               1927
22101                                        BMW_Andere               1929
22316             VW_Kaefer.__Zwei_zum_Preis_von_einem.               1000
22659                                      Opel_Corsa_B               1910
24511                       Trabant__wartburg__Ostalgie               1111
28693                                    Renault_Twingo               1910
30781                  Opel_Calibra_V6_DTM_Bausatz_1:24               1910
32585                                        UNFAL_Auto               1800
33295            Ich_verkaufe/suche_fuer_sie_ihr_Auto!!               1910
35238  Suche_Skoda_Fabia____Skoda_Fabia_Combi_mit_Klima               1500
42181         SAMSUNG_55“_3D_Tv_und_Soundbar_gegen_Auto               1910
45157                                        Motorhaube               1910
46213                                     Bellier_Vario               1910
49283                                        Citroen_HY               1001

Os registro com anos menores que 1930 não apresentam nenhum modelo que seja da época e maior que 2016 são impossíveis, devido a data de compilação do dataset

In [28]:
autos[(autos['registration_year'] < 1930) | (autos['registration_year'] > 2016)].shape[0] / autos.shape[0]

0.039631096707077984

Os dados com datas erradas representam menos de 4% de nosso dataset, sendo assim, optarei por retirar essas linhas de informação

In [29]:
autos = autos[autos["registration_year"].between(1930,2016)]

In [30]:
autos.brand.value_counts(normalize= True)

volkswagen        0.212165
bmw               0.110030
opel              0.108155
mercedes_benz     0.095386
audi              0.086428
ford              0.069764
renault           0.047349
peugeot           0.029539
fiat              0.025872
seat              0.018186
skoda             0.016040
mazda             0.015144
nissan            0.015103
smart             0.013915
citroen           0.013915
toyota            0.012478
sonstige_autos    0.010791
hyundai           0.009853
volvo             0.009249
mini              0.008645
mitsubishi        0.008145
honda             0.007853
kia               0.007103
alfa_romeo        0.006624
porsche           0.006104
suzuki            0.005916
chevrolet         0.005708
chrysler          0.003666
dacia             0.002562
daihatsu          0.002562
jeep              0.002250
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001542
daewoo            0.001500
r

Como podemos ver, as 5 principais marcas são alemãs, o que faz sentido, já que o dataset é oriundo do eBay alemão. Vamos analisar um pouco mais afundo os dados dessas montadoras 

In [31]:
topbrand = autos.brand.value_counts(normalize= True).index[:5]

In [32]:
topbrand

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi'], dtype='object')

In [33]:
autos.loc[autos['brand'].isin(topbrand)]

date_crawled                                               name  \
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
...                    ...                                                ...   
49994  2016-03-22 17:36:42  Audi_A6__S6__Avant_4.2_quattro_eventuell_Tausc...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price   abtest vehicle_type  registration_year    gearbox  power_ps  \
1       8500.0  control    limousine               1997  automatik       286   
2       8990.0     test    limousine               2009    manuell       102   
6        300.0     test    limousine               1995    manuell        90   
7       1990.0  control    limousine               1998    manuell        90   
11       350.0  control          NaN               2000        NaN         0   
...        ...      ...          ...                ...        ...       ...   
49994   5000.0  control        kombi               2001  automatik       299   
49995  24900.0  control    limousine               2011  automatik       239   
49996   1980.0  control       cabrio               1996    manuell        75   
49998  22900.0  control        kombi               2013    manuell       150   
49999   1250.0  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type          brand  \
1         7er     150000.0                   6    benzin            bmw   
2        golf      70000.0                   7    benzin     volkswagen   
6        golf     150000.0                   8    benzin     volkswagen   
7        golf     150000.0                  12    diesel     volkswagen   
11        NaN     150000.0                   0    benzin  mercedes_benz   
...       ...          ...                 ...       ...            ...   
49994      a6     150000.0                   1    benzin           audi   
49995      q5     100000.0                   1    diesel           audi   
49996   astra     150000.0                   5    benzin           opel   
49998      a3      40000.0                  11    diesel           audi   
49999  vectra     150000.0                   1    benzin           opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
6                   NaN  2016-03-20 00:00:00        31535  2016-03-23 02:48:59  
7                  nein  2016-03-16 00:00:00        53474  2016-04-07 03:17:32  
11                  NaN  2016-03-16 00:00:00        17498  2016-03-16 18:45:34  
...                 ...                  ...          ...                  ...  
49994              nein  2016-03-22 00:00:00        46537  2016-04-06 08:16:39  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[29387 rows x 17 columns]

In [34]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [35]:
mean_price = {}

for brand in topbrand:
    mean = autos['price'].loc[autos['brand'] == brand].mean()
    mileage = autos['odometer_km'].loc[autos['brand'] == brand].mean()
    mean_price[brand] = int(mean), int(mileage)

In [36]:
mean_price

{'volkswagen': (5231, 128724),
 'bmw': (8101, 132455),
 'opel': (2877, 129224),
 'mercedes_benz': (8485, 130856),
 'audi': (9093, 129287)}

In [37]:
dmp = pd.Series(mean_price)
dmp

volkswagen       (5231, 128724)
bmw              (8101, 132455)
opel             (2877, 129224)
mercedes_benz    (8485, 130856)
audi             (9093, 129287)
dtype: object

In [38]:
topbrand_df = pd.DataFrame.from_dict(data = mean_price, orient = 'index')
topbrand_df.columns = ['Average Price','Average Kms']
topbrand_df

Average Price  Average Kms
volkswagen              5231       128724
bmw                     8101       132455
opel                    2877       129224
mercedes_benz           8485       130856
audi                    9093       129287

Com isso podemos ter uma visão melhor da kilometragem e preços médios dessas marcas. Interessante notar que a média de kilometragem é similar, contudo marcas mais luxuosas como BMW, Audi e Mercedez Benz apresentam média